serveradmin/api/decorators.py

```diff
@@ -77,20 +81,133 @@ def _wrapper(request):
     return update_wrapper(_wrapper, view)
 
 
-def authenticate_app(app, token, timestamp, now, body):
-    if timestamp + 300 < now:
-        raise PermissionDenied('Expired security token')
+def authenticate_app(
+    public_keys, signatures, app_id, token, timestamp, now, body
+):
+    """Authenticate requests
 
-    real_token = calc_security_token(app.auth_token, timestamp, body)
-    if not constant_time_compare(real_token, token):
-        raise PermissionDenied('Invalid security token')
+    Ensure this request isn't beeing replayed by making sure the timestamp
+    contained in the request is no more than TIMESTAMP_GRACE_PERIOD seconds
+    removed from the current server time or raise PermissionDenied.
+
+    Hand over the real verification of auth token HMACs to authenticate_app_psk
+    or public key signatures to authenticate_app_ssh.
+
+    Ensure that the application and applications owner aren't deactivated or
+    raise PermissionDenied.
+
+    Return the app the user authenticated to
+    """
+    if (
+        timestamp + TIMESTAMP_GRACE_PERIOD < now or
```

Comment:
I think my logic makes sense:
```
In [1]: TIMESTAMP_GRACE_PERIOD = 16
   ...:
   ...: def check(now, timestamp):
   ...:     if timestamp + TIMESTAMP_GRACE_PERIOD < now:
   ...:         print('timestamp too far in the past')
   ...:     elif timestamp - TIMESTAMP_GRACE_PERIOD > now:
   ...:         print('timestamp too far in the future')
   ...:     else:
   ...:         print('timestamp okay')
   ...:
   ...: now = 1000
   ...: check(now, now)
   ...: check(now, now-100)
   ...: check(now, now+100)
   ...:
   ...:
timestamp okay
timestamp too far in the past
timestamp too far in the future
```

I added the second check as otherwise users using timestamps from the distant future generate requests which can be replayed until that time in the future has actually passed, which might be really really far away.
```
In [2]:
   ...: def old_check(now, timestamp):
   ...:     if timestamp + TIMESTAMP_GRACE_PERIOD < now:
   ...:         print('timestamp too far in the past')
   ...:     else:
   ...:         print('timestamp okay')
   ...:
   ...: old_check(now, now)
   ...: old_check(now, now-100)
   ...: old_check(now, now+100)
   ...:
   ...:
timestamp okay
timestamp too far in the past
timestamp okay
```